### ISSUE TITLE
Need to understand the results 

### ISSUE THREAD

**Issue by helio-frota** [nodeshift-starters/opossum-playground#10]
The basic-delay example has 100 samples

`lowCarb.add(delay.go, 100);`

The function fired by opossum has an delay of 100ms

The amount of  successes: 86

> Not sure if we can consider 85 since the output is based with plus 1 function call 

Why the remaining 14 are not showing as failures etc ?

```
  1 test completed.

  go x 0.00 ops/sec ±0.00% (100 runs sampled)

Function executed again. The return value is:
{
  failures: 0,
  fallbacks: 0,
  successes: 86,
  rejects: 0,
  fires: 101,
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: [Object],
  latencyTimes: [Array],
  latencyMean: 101.25581395348837
}
```

**Comment by helio-frota** [comment]
Quick reminder here that we are going back to that duplicated values 
100 runs sampled will end up with 200 fires + 1 to show the results.
but the whole idea of that 86 and the remaining 14 still the same :+1: 
our focus is on opossum results not the benchmark tooling

**Comment by alexalikiotis** [comment]
Trying doing the same test without any benchmarking framework.

```js
const CircuitBreaker = require('opossum');

const delay = (delay) => new Promise((resolve) => {
  setTimeout(() => {
    resolve();
  }, delay);
});

const breaker = new CircuitBreaker(delay);

const tests = [];

for (let i=0; i < 100; i++) {
  tests.push(breaker.fire(100));
}

Promise.all(tests).then(() => {
  console.log(breaker.stats);
});
```

Outputs the correct output
```
{
  failures: 0,
  fallbacks: 0,
  successes: 100,
  rejects: 0,
  fires: 100,
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: {
    '0': 101,
    '1': 103,
    '0.25': 101,
    '0.5': 102,
    '0.75': 102,
    '0.9': 102,
    '0.95': 103,
    '0.99': 103,
    '0.995': 103
  },
  latencyTimes: [Array],
  latencyMean: 101.76
}
```

**Comment by aalykiot** [comment]
@helio-frota 

So the problem here is inside the go function

```js
function go() {
  breaker.fire(100)
  .then()
  .catch(error => console.error(error));
  return breaker.stats;
}
``` 

So what exactly happening is that the benchmark (**lowcarb** to be exact) doesn't know how to run this function in an async style.

So what ends up happening is that the benchmark just runs the fire method of opossum thousands of times but doesn't wait for all of them to finish, now because the delay is so small (100ms) only the 86 have completed and the rest of them are still pending. 

This is why we don't see the rest 14 as failures and the output actually makes sense.

Below is the same example with just using **benchmark.js** that solves all above issues.

```js
const benchmark = require('benchmark');
const CircuitBreaker = require('opossum');

const delay = delay =>
  new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, delay);
  });

const breaker = new CircuitBreaker(delay);
const suite = new benchmark.Suite();

suite.add('opossum-delay', {
  defer: true,
  minSamples: 100,
  fn: deferred => {
    breaker.fire(100).then(() => {
      deferred.resolve();
    });
  },
});

suite.on('complete', () => {
  console.log(breaker.stats);
});

suite.run();

**Comment by helio-frota** [comment]
> So what exactly happening is that the benchmark (lowcarb to be exact) doesn't know how to run this function in an async style.

hm. I'm not sure about this because the third parameter here: https://github.com/nodeshift-starters/opossum-playground/blob/master/basic-delay/benchmark/index.js#L6 is to run benchmark.js in async mode.

> Below is the same example with just using benchmark.js that solves all above issues.

So when executing your example , we still have the situation described in the issue:

```
{
  failures: 0,
  fallbacks: 0,
  successes: 79,   <---- 
  rejects: 0,
  fires: 78,     <------
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
```

What happened with the 21 other ticks ? Why it is not showing as failures or rejects or etc ?
This is exactly the goal of this issue. To understand were are / what happened with the other values. :+1: 

Btw I'm going to update this example but the current output is:

```
  1 test completed.

  go x 16,362 ops/sec ±5.27% (100 runs sampled)

Function executed again. The return value is:
{
  failures: 0,
  fallbacks: 0,
  successes: 173,   <--------
  rejects: 0,
  fires: 201,  <-------- 2x 100 runs sampled +  1 to show the results
```

So we have `28` (201-173) executions without any failures, fallbacks, rejects, success, etc...

**Comment by helio-frota** [comment]
@lance hey, question about opossum. 

For each "fire" count, opossum needs to have respective stats to sum the amount of "fire"s ?

For example:

```
{
  failures: 0,
  fallbacks: 0,
  successes: 10,   <-------- should be 10 success
  rejects: 0,
  fires: 10   <---- 10 fires 
}
```

and for example (7 success +  2 failures + 1 reject === 10 fires)

```
{
  failures: 2,        <--------
  fallbacks: 0,
  successes: 7,   <--------
  rejects: 1,          <--------
  fires: 10   <---- 10 fires 
}
```

Thanks

**Comment by aalykiot** [comment]
@helio-frota 

> hm. I'm not sure about this because the third parameter here: https://github.com/nodeshift-starters/opossum-playground/blob/master/basic-delay/benchmark/index.js#L6 is to run benchmark.js in async mode.

The async parameter that we pass in the benchmark's run function means that we want the benchmark suite to run asynchronously (not blocking the event loop) relative to the underneath code. (https://benchmarkjs.com/docs#options_async)

But the function we provide to the benchmark will still run synchronously whether we set the async attribute to true or false.

For async executions in each iteration of our function in the suite check out the following on stackoverflow: https://stackoverflow.com/questions/31624055/benchmark-asynchronous-code-benchmark-js-node-js/31624409#31624409 

> Btw I'm going to update this example but the current output is:
> 
>
>   1 test completed.
> 
>   go x 16,362 ops/sec ±5.27% (100 runs sampled)
> 
> Function executed again. The return value is:
> {
>   failures: 0,
>   fallbacks: 0,
>   successes: 173,   <--------
>   rejects: 0,
>   fires: 201,  <-------- 2x 100 runs sampled +  1 to show the results
> 
> So we have 28 (201-173) executions without any failures, fallbacks, rejects, success, etc...

This is still happening because you run the go function synchronously inside the benchmark suite. You are not missing anything is just the rest 28 haven't completed yet. (and you always get 28 pending because the delay on each fire is constant)

> So when executing your example, we still have the situation described in the issue:
> 
> {
>   failures: 0,
>   fallbacks: 0,
>   successes: 79,   <---- 
>   rejects: 0,
>   fires: 78,     <------
>   timeouts: 0,
>   cacheHits: 0,
>   cacheMisses: 0,
>   semaphoreRejections: 0,

I still believe that we're getting this (kind of) weird behavior because of the way we use the benchamrk.js tool (still I'm not very familiar with it and how to use it properly) rather than the opossum has a bug or something.

Though, I'll continue to play around with this example and try to find why we're getting this result. ✌️

**Comment by helio-frota** [comment]
> But the function we provide to the benchmark will still run synchronously whether we set the async attribute to true or false.

gotcha 

> I still believe that we're getting this (kind of) weird behavior because of the way we use the benchamrk.js tool (still I'm not very familiar with it and how to use it properly) rather than the opossum has a bug or something.

Yeah that may be. Also we don't need to worry about to find bugs on opossum. (not atm) :+1:

**Comment by helio-frota** [comment]
> I still believe that we're getting this (kind of) weird behavior because of the way we use the benchamrk.js tool (still I'm not very familiar with it and how to use it properly) rather than the opossum has a bug or something.

Yeah that seems correct, probably we should try other benchmark 

This example is using `ab` command and the output is normal https://github.com/nodeshift-starters/opossum-playground/tree/master/two-express (also this example is based on real use case)

So yeah I think we should investigate why that happen when using bench tools

**Comment by helio-frota** [comment]
@alexalikiotis @lance I start to think that we are in good shape and the underline bench tools usage is not interfering... but feel free to add comments on this:

Using the `basic-delay` example:

When changing the the delay from 100ms to 1ms

```
function go() {
  breaker
    .fire(1) < ---- the delay in ms ----
    .then()
```

We have almost the same numbers of fires and successes:

```
{
  failures: 0,
  fallbacks: 0,
  successes: 198,  < -----
  rejects: 0,
  fires: 201,  < ------
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: [Object],
  latencyTimes: [Array],
  latencyMean: 1.9494949494949494
}
```

And `latencyMean` 1.9+.

Also when increasing the delay to `500` we have less successes:

```
{
  failures: 0,
  fallbacks: 0,
  successes: 72,  < -----
  rejects: 0,
  fires: 201,  < -----
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: [Object],
  latencyTimes: [Array],
  latencyMean: 500.7361111111111
}
```

And `latencyMean` 500+.

And this is an interesting part:

When changing back the delay to `100` and setting the opossum timeout option to `102` 

```
const options = {
  timeout: 102
};

function go() {
  breaker
    .fire(100)
    .then()
```


```
{
  failures: 1,          < --------
  fallbacks: 0,
  successes: 173,   < ------
  rejects: 0,
  fires: 201,         <--------
  timeouts: 1,       < -------
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: [Object],
  latencyTimes: [Array],
  latencyMean: 100.92571428571429  < ----
}
```

We have 1 failure, which is a (1) timeout , a latencyMean 100+

Even it still a little bit complicated for me to explain, I think opossum is working properly but I just want to know why
`successes` !=== `fires` ,  @lance is the `latencyMean` the "valid delay" that blocks `successes` to be increased ?

In other words: The "Latency mean" is the delay-register-of-the-cause of not updating successes, and this is expected. (which makes sense)

**Comment by lance** [comment]
@helio-frota the problem is with how the async function execution is being handled. In the function being benchmarked, when you return the stats, not all executions have completed. In #15 I changed the return value so that it's returning a `Promise` which will eventually resolve with the correct results. This change, however, depends on https://github.com/helio-frota/lowcarb/pull/6. In `lowcarb` the test suite needs to wait for these promises to resolve. Async is tricky!

With these changes, I see this output.

```log
opossum-playground/basic-delay on  master [!] is 📦 v1.0.0 via ⬢ v12.16.3 
❯ npm start                                                                                   1755ms  Tue 21 Jul 2020 09:10:04 PM EDT

> basic-delay@1.0.0 start /home/lanceball/src/github.com/opossum-playground/basic-delay
> lowcarb


  1 test completed.

  go x 12,245 ops/sec ±8.87% (100 runs sampled)

Function executed again. The return value is:
{
  failures: 0,
  fallbacks: 0,
  successes: 201,
  rejects: 0,
  fires: 201,
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: [Object],
  latencyTimes: [Array],
  latencyMean: 105.53233830845771
}
```

**Comment by helio-frota** [comment]
1) PR on lowcarb merged and  0.5.0 published
2) PR of the example merged 
3) :tada: 

```
  1 test completed.

  go x 38,684 ops/sec ±5.14% (100 runs sampled)

Function executed again. The return value is:
{
  failures: 0,
  fallbacks: 0,
  successes: 201,
  rejects: 0,
  fires: 201,
  timeouts: 0,
  cacheHits: 0,
  cacheMisses: 0,
  semaphoreRejections: 0,
  percentiles: [Object],
  latencyTimes: [Array],
  latencyMean: 101.19402985074628
}
```

@lance @alexalikiotis Thanks!
